### PR TITLE
Add AWS S3 session token config

### PR DESCRIPTION
**Pull request by tinkerrrr**
Add AWS S3 session token config to support `BasicSessionCredentials`

### CODE REVIEW

**Review by Praveen2112 — APPROVED** [review]
Thanks for working on this.

**Comment by Praveen2112** [comment]
I have one follow-up question, won't these session token have a lifetime ?

**Comment by tinkerrrr** [comment]
> I have one follow-up question, won't these session token have a lifetime ?

As for as I know, yes, if you send requests using expired credentials, Amazon S3 will return an error. I need this because when I run presto on S3 with only temporary credentials (like an assume-role token), there should be a place to configure it.

**Comment by Praveen2112** [comment]
@tinkerrrr Can you please rebase this PR ?

File: presto-hive/src/main/java/io/prestosql/plugin/hive/s3/S3SecurityMapping.java [hunk 1]
```diff
@@ -72,8 +75,15 @@ public S3SecurityMapping(
 
         requireNonNull(accessKey, "accessKey is null");
         requireNonNull(secretKey, "secretKey is null");
+        requireNonNull(sessionToken, "sessionToken is null");
         checkArgument(accessKey.isPresent() == secretKey.isPresent(), "accessKey and secretKey must be provided together");
-        this.credentials = accessKey.map(access -> new BasicAWSCredentials(access, secretKey.get()));
+
+        if (sessionToken.isEmpty()) {
```
Comment (Praveen2112) [inline presto-hive/src/main/java/io/prestosql/plugin/hive/s3/S3SecurityMapping.java:81]: Can we swap the conditions ?

**Comment by tinkerrrr** [comment]
> @tinkerrrr Can you please rebase this PR ?

done

**Comment by Praveen2112** [comment]
From AWS docs for reference - 
> The duration of temporary credentials can range from 900 seconds (15 minutes) to 129600 seconds (36 hours) for IAM users. If a duration isn’t specified, then 43200 seconds (12 hours) is used by default.

**Comment by Praveen2112** [comment]
@tinkerrrr Sorry to bother you. Can you rebase the PR to resolve the merge conflicts

**Comment by tinkerrrr** [comment]
@Praveen2112 Resolved

**Comment by findepi** [comment]
@tinkerrrr can you please rebase your PR so that there are no merge commits?

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMapping.java [hunk 2]
```diff
@@ -84,7 +87,9 @@ public S3SecurityMapping(
         requireNonNull(accessKey, "accessKey is null");
         requireNonNull(secretKey, "secretKey is null");
         checkArgument(accessKey.isPresent() == secretKey.isPresent(), "accessKey and secretKey must be provided together");
-        this.credentials = accessKey.map(access -> new BasicAWSCredentials(access, secretKey.get()));
+        this.credentials = sessionToken.isEmpty() ?
```
Comment (losipiuk) [inline plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMapping.java:90]: validate that if `sessionToken` is provided then we require also `accessKey` and `secretKey` to be provided.

**Review by electrum — CHANGES_REQUESTED** [review]
Can you explain the use case for this feature? My understanding, based on the [AWS documentation](https://docs.aws.amazon.com/STS/latest/APIReference/API_GetSessionToken.html), is that these are for temporary credentials:

> Returns a set of temporary credentials for an AWS account or IAM user.

> This duration can range from ... up to a maximum of [36 hours] with a default of [12 hours]

This means anyone using the feature would have a cluster that stops working after a period of time. Instead, you can use the IAM role feature, which will cause the S3 client to use STS to assume the role.

**Comment by tinkerrrr** [comment]
@electrum The use case is, when deploy Presto cluster on AWS, in the testing phase, it's unsafe to test with formal access key and secret key or full-authority assume roles, so sessionToken with expire time could resolve it.

File: docs/src/main/sphinx/connector/hive-s3.rst [hunk 3]
```diff
@@ -147,6 +149,10 @@ The security mapping must provide one or more configuration settings:
 * ``accessKey`` and ``secretKey``: AWS access key and secret key. This overrides
   any globally configured credentials, such as access key or instance credentials.
 
+* ``sessionToken``: AWS credential session token. This configuration wont't take effect
```
[the remark this answers is not in the document]
Comment (tinkerrrr) [reply docs/src/main/sphinx/connector/hive-s3.rst]: done

File: plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMapping.java (hunk 2 above)
Comment (tinkerrrr) [reply plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/S3SecurityMapping.java:93]: done

**Comment by losipiuk** [comment]
CI: https://github.com/trinodb/trino/issues/6723

**Review by losipiuk — APPROVED** [review]
@electrum Are you good with this one?

**Comment by alexjo2144** [comment]
Rather than specify the token in a config file like this, you could pass them via environment variables or system properties. Or have it written to the `~/.aws/credentials` file. The AWS credential provider chain should pick any of those up.

**Comment by tinkerrrr** [comment]
@alexjo2144 I think they are different cases, using environment variables, system properties or `~/.aws/credentials` file is kind of unsafe, since other process can use them too. Meanwhile, from https://github.com/trinodb/trino/blob/master/plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java#L968, it seems trino tends to provide `BasicSessionCredentials ` from session token config.

@electrum @losipiuk How do you think?

**Comment by losipiuk** [comment]
> This means anyone using the feature would have a cluster that stops working after a period of time. Instead, you can use the IAM role feature, which will cause the S3 client to use STS to assume the role.

It would work if the value is changed in config on regular basis and automatic reloading of security mapping file is enabled, rigth?

> @alexjo2144 I think they are different cases, using environment variables, system properties or `~/.aws/credentials` file is kind of unsafe, since other process can use them too. Meanwhile, from https://github.com/trinodb/trino/blob/master/plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java#L968, it seems trino tends to provide `BasicSessionCredentials ` from session token config.
> 
> @electrum @losipiuk How do you think?

I do not feel strongly. Though I would prefer to not extend product surface solely for test-only uscase. @tinkerrrr Does approach suggested by @electrum work for you?

**Comment by tinkerrrr** [comment]
@losipiuk For some reason, all I got are session token and corresponding AKSK in the beginning, so there's no IAM role to configure. Anyway, I prefer to add this config since `BasicSessionCredentials` is provided in `TrinoS3FileSystem`

**Comment by electrum** [comment]
@tinkerrrr Sorry, I'm still having trouble understanding why this is preferable to IAM roles for EC2 instances. Everything I can find in the AWS documentation says that best practice is to use IAM roles for temporary credentials for EC2 instances:

[Security best practices in IAM](https://docs.aws.amazon.com/IAM/latest/UserGuide/best-practices.html#use-roles-with-ec2)

> **Use roles for applications that run on Amazon EC2 instances**
> Applications that run on an Amazon EC2 instance need credentials in order to access other AWS services. To provide credentials to the application in a secure way, use IAM roles. ... In the case of Amazon EC2, IAM dynamically provides temporary credentials to the EC2 instance, and these credentials are automatically rotated for you.

[Temporary security credentials in IAM](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp.html)

> **Common scenarios for temporary credentials**
> **Roles for Amazon EC2**
> If you run applications on Amazon EC2 instances and those applications need access to AWS resources, you can provide temporary security credentials to your instances when you launch them. These temporary security credentials are available to all applications that run on the instance, so you don't need to store any long-term credentials on the instance. For more information, see *Using an IAM role to grant permissions to applications running on Amazon EC2 instances*.

The support for session token in `TrinoS3FileSystem` was added for programmatic use, not intended to be user configurable.

**Comment by tinkerrrr** [comment]
@electrum It's a test-only use case came from a special scenario, so it doesn't follow the best practice from AWS documentation... 

If session token is not user configurable, the following code is never reached

```java
String sessionToken = conf.get(S3_SESSION_TOKEN);
if (!isNullOrEmpty(sessionToken)) {
    return Optional.of(new BasicSessionCredentials(accessKey, secretKey, sessionToken));
}
```

I think the point is:

> Though I would prefer to not extend product surface solely for test-only uscase.

If it's considered as not appropriate, then I'll close this PR

**Pull request closed without merging**